core/src/main/scala/org/http4s/Message.scala

```diff
@@ -55,328 +30,157 @@ sealed trait Message[F[_]] extends MessageOps[F] { self =>
-  * @param attributes Immutable Map used for carrying additional information in a type safe fashion
-  */
-sealed abstract case class Request[F[_]](
-    method: Method = Method.GET,
-    uri: Uri = Uri(path = "/"),
-    httpVersion: HttpVersion = HttpVersion.`HTTP/1.1`,
-    headers: Headers = Headers.empty,
-    body: EntityBody[F] = EmptyBody,
-    attributes: AttributeMap = AttributeMap.empty
-) extends Message[F]
-    with RequestOps[F] {
-  import Request._
-
-  type Self = Request[F]
-
-  private def requestCopy(
-      method: Method = this.method,
-      uri: Uri = this.uri,
-      httpVersion: HttpVersion = this.httpVersion,
-      headers: Headers = this.headers,
-      body: EntityBody[F] = this.body,
-      attributes: AttributeMap = this.attributes
-  ): Request[F] =
-    Request(
-      method = method,
-      uri = uri,
-      httpVersion = httpVersion,
-      headers = headers,
-      body = body,
-      attributes = attributes
-    )
-
-  @deprecated(
-    message = "Copy method is unsafe for setting path info. Use with... methods instead",
-    "0.17.0-M3")
-  def copy(
-      method: Method = this.method,
-      uri: Uri = this.uri,
-      httpVersion: HttpVersion = this.httpVersion,
-      headers: Headers = this.headers,
-      body: EntityBody[F] = this.body,
-      attributes: AttributeMap = this.attributes
-  ): Request[F] =
-    requestCopy(
-      method = method,
-      uri = uri,
-      httpVersion = httpVersion,
-      headers = headers,
-      body = body,
-      attributes = attributes
-    )
-
-  def withMethod(method: Method) = requestCopy(method = method)
-  def withUri(uri: Uri) =
-    requestCopy(uri = uri, attributes = attributes -- Request.Keys.PathInfoCaret)
-  def withHttpVersion(httpVersion: HttpVersion) = requestCopy(httpVersion = httpVersion)
-  def withHeaders(headers: Headers) = requestCopy(headers = headers)
-  def withAttributes(attributes: AttributeMap) = requestCopy(attributes = attributes)
-
-  def withBodyStream(body: EntityBody[F]): Request[F] =
-    requestCopy(body = body)
-
-  override protected def change(
-      body: EntityBody[F],
-      headers: Headers,
-      attributes: AttributeMap): Self =
-    requestCopy(body = body, headers = headers, attributes = attributes)
-
-  lazy val authType: Option[AuthScheme] = headers.get(Authorization).map(_.credentials.authScheme)
-
-  lazy val (scriptName, pathInfo) = {
-    val caret = attributes.get(Request.Keys.PathInfoCaret).getOrElse(0)
-    uri.path.splitAt(caret)
-  }
+  def transformHeaders(m: M[F])(f: Headers => Headers): M[F] =
+    change(m)(body(m), f(headers(m)), attributes(m))
 
-  def withPathInfo(pi: String)(implicit F: Functor[F]): Self =
-    withUri(uri.withPath(scriptName + pi))
+  def replaceAllHeaders(m: M[F])(headers: Headers): M[F] =
+    change(m)(body(m), headers, attributes(m))
 
-  lazy val pathTranslated: Option[File] = attributes.get(Keys.PathTranslated)
+  def replaceAllHeadersWith(m: M[F])(headers: Header*): M[F] =
```

Comment:
What prompted this name change?  This seemed to ripple hard in the tests.

Reply:
Name overloading did not work on implicit classes.